### PR TITLE
Export/Import Properties of CorrelationFilter

### PR DESCRIPTION
**Pull request by robionekenobi**
In the current version, when exporting topics, the eventual properties of the CorrelationFilter of a subscription is not exported, nor imporetd
This pull requesrt correct this

### CODE REVIEW

File: src/Common/Helpers/ImportExportHelper.cs [hunk 1]
```diff
@@ -336,8 +339,13 @@ void GetProperties(Type type, bool canRead, bool canWrite)
             {
                 return;
             }
+            //
+            // We must add the Properties of the CorrelationFilter. As the Properties are not CanWrite, we must specifically add it by not testing the canWrite parameter
+            //
```
Comment (SeanFeldman) [inline src/Common/Helpers/ImportExportHelper.cs:344]:
```suggestion

            // We must add the Properties of the CorrelationFilter. As the Properties are not CanWrite, we must specifically add it by not testing the canWrite parameter
```

File: src/Common/Helpers/ImportExportHelper.cs [hunk 2]
```diff
@@ -423,6 +431,26 @@ private static void GetPropertyValue(Dictionary<string, PropertyInfo> propertyDi
             if (property.PropertyType == typeof(string))
             {
                 propertyValue[name] = xmlReader.ReadElementContentAsString();
+                return;
+            }
+            //
+            // Here we fill in the Property of the Filter CorrelationId Properties by looping through the Property Sibling
+            //
```
Comment (SeanFeldman) [inline src/Common/Helpers/ImportExportHelper.cs:438]:
```suggestion

            // Here we fill in the Property of the Filter CorrelationId Properties by looping through the Property Sibling
```

File: src/Common/Helpers/ImportExportHelper.cs [hunk 3]
```diff
@@ -423,6 +431,26 @@ private static void GetPropertyValue(Dictionary<string, PropertyInfo> propertyDi
...
+            if (property.MemberType == MemberTypes.Property && property.PropertyType.Name == "IDictionary`2")
+            {
+                bool bOk = xmlReader.ReadToDescendant("Property");
+                while (bOk)
+                {
+                    var _read = xmlReader.ReadElementContentAsString();
```
Comment (SeanFeldman) [inline src/Common/Helpers/ImportExportHelper.cs:444]: Local variable don't have `_` in this code base. The underscore is reserved for member fields. Could you please rename? Thank you.

File: src/Common/Helpers/ImportExportHelper.cs [hunk 4]
```diff
@@ -423,6 +431,26 @@ private static void GetPropertyValue(Dictionary<string, PropertyInfo> propertyDi
...
+                    var _prop = new Property();
+                    _prop.Name = _read.Split(':')[0];
+                    _prop.Value = _read.Split(':')[1];
```
Comment (SeanFeldman) [inline src/Common/Helpers/ImportExportHelper.cs:447]: The split doesn't need to be happening each time. Split once and use that array parts.

File: src/Common/Helpers/ImportExportHelper.cs [hunk 5]
```diff
@@ -533,6 +561,24 @@ async Task SerializeEntity(ServiceBusHelper serviceBusHelper, XmlWriter xmlWrite
                     xmlWriter.WriteEndElement();
                     continue;
                 }
+                //
+                // We check if the CorrelationFilter has Properties.
+                // If yes we write the xml <Properties><Property>A</Property></Properties>
+                //
```
Comment (SeanFeldman) [inline src/Common/Helpers/ImportExportHelper.cs:567]:
```suggestion

                // We check if the CorrelationFilter has Properties.
                // If yes we write the xml <Properties><Property>A</Property></Properties>
```

File: src/Common/Helpers/ImportExportHelper.cs [hunk 6]
```diff
@@ -1349,11 +1395,30 @@ private Filter CreateFilter(XElement filter)
             if (filter.Name == string.Format(NodeNameFormat, Namespace, CorrelationFilterEntity))
             {
                 ruleFilter = new CorrelationFilter(propertyValue[CorrelationId] as string);
+                //
+                // We check if there is a Key with Properties
+                // If Yes, we must create the properties of the CorrelationFilter
+                //
```
Comment (SeanFeldman) [inline src/Common/Helpers/ImportExportHelper.cs:1401]:
```suggestion

                // We check if there is a Key with Properties
                // If Yes, we must create the properties of the CorrelationFilter
```

File: src/Common/Helpers/ImportExportHelper.cs [hunk 7]
```diff
@@ -1349,11 +1395,30 @@ private Filter CreateFilter(XElement filter)
...
+                        _rule.Properties.Add(_prop.Name, _prop.Value);
+                    }
+
+                }
             }
             if (ruleFilter != null)
             {
+                //
+                // As we already created the Properties, we must remove it here before calling SetPropertyValue, as Properties Property is ReadOnly
+                //
```
Comment (SeanFeldman) [inline src/Common/Helpers/ImportExportHelper.cs:1417]:
```suggestion

                // As we already created the Properties, we must remove it here before calling SetPropertyValue, as Properties Property is ReadOnly
```

**Review by SeanFeldman — CHANGES_REQUESTED** [review]
Thank you for the PR, @robionekenobi.
A few minor updates if you don't mind.

File: src/Common/Helpers/ImportExportHelper.cs (hunk 3 above)
Comment (robionekenobi) [reply src/Common/Helpers/ImportExportHelper.cs]: Hi, corrected and improved

File: src/Common/Helpers/ImportExportHelper.cs (hunk 4 above)
Comment (robionekenobi) [reply src/Common/Helpers/ImportExportHelper.cs]: Hi, corrected and improved

File: src/Common/Helpers/ImportExportHelper.cs [hunk 8]
```diff
@@ -1349,11 +1392,28 @@ private Filter CreateFilter(XElement filter)
             if (filter.Name == string.Format(NodeNameFormat, Namespace, CorrelationFilterEntity))
             {
                 ruleFilter = new CorrelationFilter(propertyValue[CorrelationId] as string);
+
+                // We check if there is a Key with Properties
+                // If Yes, we must create the properties of the CorrelationFilter
+                if (propertyValue.ContainsKey(CorrelationFilterProperties))
+                {
+                    var _rule = ruleFilter as CorrelationFilter;
```
Comment (SeanFeldman) [inline src/Common/Helpers/ImportExportHelper.cs:1400]: To be consistent with the project, could you remove `_` from local variables?

File: src/Common/Helpers/ImportExportHelper.cs [hunk 9]
```diff
@@ -1349,11 +1392,28 @@ private Filter CreateFilter(XElement filter)
...
+                        _rule.Properties.Add(_prop.Name, _prop.Value);
+                    }
+
+                }
             }
             if (ruleFilter != null)
             {
+
+                // As we already created the Properties, we must remove it here before calling SetPropertyValue, as Properties Property is ReadOnly
+                var _propVal = propertyValue.Where(p => p.Key != CorrelationFilterProperties);
```
Comment (SeanFeldman) [inline src/Common/Helpers/ImportExportHelper.cs:1413]: Here as well.

**Review by SeanFeldman — COMMENTED** [review]
Two more spots and you're good 🙂

File: src/Common/Helpers/ImportExportHelper.cs (hunk 8 above)
Comment (robionekenobi) [reply src/Common/Helpers/ImportExportHelper.cs]: Done

File: src/Common/Helpers/ImportExportHelper.cs (hunk 9 above)
Comment (robionekenobi) [reply src/Common/Helpers/ImportExportHelper.cs]: Done

**Review by SeanFeldman — APPROVED** [review]

**Review by ErikMogensen — APPROVED** [review]
Nice

**Pull request merged**